osquery/distributed/distributed.cpp

```diff
@@ -80,73 +100,173 @@ Status Distributed::pullUpdates() {
   return Status(0, "OK");
 }
 
+/**
+ * getPendingQueryCount
+ * Return number of queries waiting to be executed.
+ */
 size_t Distributed::getPendingQueryCount() {
-  std::vector<std::string> queries;
-  scanDatabaseKeys(kQueries, queries, kDistributedQueryPrefix);
-  return queries.size();
+  int num = 0;
+  for (auto item : results_) {
+    if (item.isPending()) {
+      num += 1;
+    }
+  }
+  return num;
 }
 
-size_t Distributed::getCompletedCount() {
-  return results_.size();
+// Returns the number of time distributed_read endpoint was accessed
+size_t Distributed::numDistReads() {
+  return numDistReads_;
 }
 
-Status Distributed::serializeResults(std::string& json) {
+// Returns the number of time distributed_write endpoint was accessed
+size_t Distributed::numDistWrites() {
+  return numDistWrites_;
+}
+
+Status Distributed::serializeResults(std::string& json,
+                                     DistributedQueryResult* item) {
   auto doc = JSON::newObject();
   auto queries_obj = doc.getObject();
   auto statuses_obj = doc.getObject();
   for (const auto& result : results_) {
+    if (result.isPending() || result.hasReported) {
+      continue;
+    }
+    if (0L != item && &result != item) {
+      continue;
+    }
     auto arr = doc.getArray();
     auto s = serializeQueryData(result.results, result.columns, doc, arr);
     if (!s.ok()) {
       return s;
     }
-    doc.add(result.request.id, arr, queries_obj);
-    doc.add(result.request.id, result.status.getCode(), statuses_obj);
+    doc.add(result.id, arr, queries_obj);
+    doc.add(result.id, result.status.getCode(), statuses_obj);
   }
 
   doc.add("queries", queries_obj);
   doc.add("statuses", statuses_obj);
   return doc.toString(json);
 }
 
-void Distributed::addResult(const DistributedQueryResult& result) {
-  results_.push_back(result);
-}
-
 Status Distributed::runQueries() {
-  while (getPendingQueryCount() > 0) {
-    auto request = popRequest();
-    LOG(INFO) << "Executing distributed query: " << request.id << ": "
-              << request.query;
+  // sanity check - make sure plugin available
+  // Shouldn't ever fail, since read endpoint worked, so plugin must be active
+  auto distributed_plugin = RegistryFactory::get().getActive("distributed");
+  if (!RegistryFactory::get().exists("distributed", distributed_plugin)) {
+    return Status(1, "Missing distributed plugin " + distributed_plugin);
+  }
+
+  int i = -1;
+  for (auto& item : results_) {
+    i++;
+    if (false == item.isPending()) {
+      continue;
+    }
 
-    // Keep track of the currently executing request
-    Distributed::setCurrentRequestId(request.id);
+    // if a distributed query has a long list, may need a
+    // delay in between to avoid being killed by watcher.
+    if (i > 0 && FLAGS_distributed_intra_sleep > 0) {
+      if (FLAGS_distributed_intra_sleep > 30) {
```

Comment:
Should you notify the user you're overriding their flag settings here?

Reply:
Added LOG line